spd/models/components.py

```diff
@@ -161,10 +160,68 @@ def forward(self, x: Float[Tensor, "batch pos"]) -> Float[Tensor, "batch pos emb
         # From https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L1211
         component_acts = self.V[x]  # (batch pos C)
 
-        if self.mask is not None:
-            component_acts *= self.mask
+        if mask_BxC is not None:
+            component_acts *= mask_BxC
 
         out = einops.einsum(
             component_acts, self.U, "batch pos C, ... C embedding_dim -> batch pos embedding_dim"
         )
         return out
+
+
+# TODO(oli) make this the only public class here
+class ReplacedComponent(nn.Module):
+    def __init__(
+        self,
+        original: nn.Linear | nn.Embedding,
+        replacement: LinearComponent | EmbeddingComponent,
+    ):
+        super().__init__()
+        assert isinstance(original, nn.Linear) == isinstance(replacement, LinearComponent)
+        self.original = original
+        self.replacement = replacement
+
+        self.forward_mode: Literal["original"] | Literal["replacement"] | None = None
+        self.mask_BxC: Tensor | None = None
+        self.init_weights_()
+
+    def init_weights_(self) -> None:
```

Comment:
I don't love that the init logic is here. This class feels like its job is just to choose which module to run through. Adding init logic seems out of place. I think one of the following might be nicer (roughly in order of niceness IMO):
1. Call a new init_weights function from both LinearComponent and EmbeddingComponent.
2. Create something like a Component class (from which LinearComponent and EmbeddingComponent inherits from), and that class this init function.
3. Just calling this init function after initialization of the class. Or maybe at the end of the ComponentModel init.

Thoughts?